include/sundials/impl/sundials_errors_impl.h

```diff
@@ -0,0 +1,183 @@
+                   SUNCTX);                                              \
+      return NULL;                                                       \
+    }                                                                    \
+  }                                                                      \
+  while (0)
+#else
+#define SUNCheckCallNull(call) (void)call;
+#endif
+
+/* Same as SUNCheckCall, but returns void. */
+#if !defined(SUNDIALS_DISABLE_ERROR_CHECKS)
+#define SUNCheckCallVoid(call)                                           \
+  do {                                                                   \
+    SUNErrCode sun_chk_call_err_code_ = call;                            \
+    if (SUNHintFalse(sun_chk_call_err_code_ < 0))                        \
+    {                                                                    \
+      SUNHandleErr(__LINE__, __func__, __FILE__, sun_chk_call_err_code_, \
+                   SUNCTX);                                              \
+      return;                                                            \
+    }                                                                    \
+  }                                                                      \
+  while (0)
+#else
+#define SUNCheckCallNull(call) (void)call;
+#endif
+
+/* SUNCheckLastErr checks the last_err value in the SUNContext.
+   If an error occured, then it will log the error, set the last_err
+   value, and calls the error handler. */
+
+#if !defined(SUNDIALS_DISABLE_ERROR_CHECKS)
+#define SUNCheckCallLastErrNoRet(call) \
+  call;                                \
+  SUNCheckCallNoRet(SUNGetLastErr(SUNCTX))
+
+/* Same as SUNCheckCallLastErrNoRet, but returns with the error code. */
+#define SUNCheckCallLastErr(call) \
+  call;                           \
+  SUNCheckCall(SUNGetLastErr(SUNCTX))
+
+/* Same as SUNCheckCallLastErrNoRet, but returns void. */
+#define SUNCheckCallLastErrVoid(call) \
+  call;                               \
+  SUNCheckCallVoid(SUNGetLastErr(SUNCTX))
+
+/* Same as SUNCheckCallLastErrNoRet, but returns with the error code. */
+#define SUNCheckCallLastErrNull(call) \
+  call;                               \
+  SUNCheckCallNull(SUNGetLastErr(SUNCTX))
+#else
+#define SUNCheckCallLastErrNoRet(call) call
+#define SUNCheckCallLastErr(call)      call
+#define SUNCheckCallLastErrVoid(call)  call
+#define SUNCheckCallLastErrNull(call)  call
+#endif
+
+/* SUNAssert checks if an expression is true.
+   If the expression is false, it calls the SUNAssertErrHandler. */
+#if !defined(NDEBUG)
+#define SUNAssert(expr, code)                                          \
+  do {                                                                 \
+    if (!(expr))                                                       \
+    {                                                                  \
+      SUNAssertErrHandlerFn(__LINE__, __func__, __FILE__, #expr, code, \
```

Comment:
I think it would be expected from "assert" but it might be good to note in the documentation that this does not go through the error handler stack like the other macros but goes directly to the SUNDIALS assert error handler function.